presto-main/src/main/java/com/facebook/presto/server/protocol/ExecuteAndOutputHandler.java

```diff
@@ -0,0 +1,117 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.server.protocol;
+
+import com.facebook.presto.Session;
+import com.facebook.presto.client.Column;
+import com.facebook.presto.common.QueryTypeAndExecutionExtraMessage;
+import com.facebook.presto.common.QueryTypeAndExecutionExtraMessage.ExecutionExtraMessage;
+import com.facebook.presto.common.resourceGroups.QueryType;
+import com.facebook.presto.common.type.Type;
+import com.facebook.presto.metadata.Metadata;
+import com.facebook.presto.sql.analyzer.BuiltInQueryAnalysis;
+import com.facebook.presto.sql.analyzer.utils.StatementUtils;
+import com.facebook.presto.sql.tree.Statement;
+import com.google.common.collect.ImmutableList;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Consumer;
+import java.util.function.Predicate;
+
+import static java.util.Objects.requireNonNull;
+
+public abstract class ExecuteAndOutputHandler<T extends ExecutionExtraMessage>
+{
+    // intermediary information calculated and set during execution, and maybe used in order to refactor output
+    QueryTypeAndExecutionExtraMessage<T> queryExtraMessage;
+
+    public void executeInTask(Statement statement, Optional<BuiltInQueryAnalysis> queryAnalysis, Metadata metadata, Session session)
+    {
+        String updateType = StatementUtils.getQueryType(statement.getClass()).map(QueryType::toString)
+                .orElse("UNKNOWN");
+        this.queryExtraMessage = new QueryTypeAndExecutionExtraMessage(updateType, execute(statement, queryAnalysis, metadata, session));
+    }
+
+    public OutputResultData handleOutputWithResult(List<Column> columns, Iterable<List<Object>> value,
+                                                   Predicate predicate, Consumer<Long> consumer)
+    {
+        return OutputResultData.EMPTY;
+    }
+
+    public OutputColumn getOutputColumns()
+    {
+        return OutputColumn.EMPTY;
+    }
+
+    public OutputResultData handleOutputWithoutResult()
+    {
+        return OutputResultData.EMPTY;
+    }
+
+    abstract T execute(Statement statement, Optional<BuiltInQueryAnalysis> queryAnalysis, Metadata metadata, Session session);
+
+    public static class OutputResultData
+    {
+        public static final OutputResultData EMPTY = new OutputResultData(false, ImmutableList.of());
+        private final boolean needRefactor;
+        private final Iterable<List<Object>> value;
+
+        public OutputResultData(boolean needRefactor, Iterable<List<Object>> value)
+        {
+            this.needRefactor = needRefactor;
+            this.value = requireNonNull(value, "value is null");
+        }
+
+        public boolean isNeedRefactor()
+        {
+            return needRefactor;
+        }
```

Comment:
Why is this called `needRefactor`?

Reply:
Oh, maybe an inaccurate wording. What I mean is if need to overwrite the original columns or result data, so I have renamed it to ``needOverwrite``. Is it OK?